### PR TITLE
[8.x] Add `takeUntilTimeout` method to the lazy collection

### PR DESCRIPTION
**Pull request by JosephSilber**
This PR adds a `takeUntilTimeout` to the `LazyCollection` class, which lets you stop a long running process after a given timeout:

```php
$lazyCollection
    ->takeUntilTimeout(now()->add(2, 'minutes'))
    ->each(fn ($item) => doSomethingThatMayTakeSomeTime($item));

// ^^ This will only process items for up to 2 minutes ^^
```

---

This is **especially useful for Laravel Vapor apps**, since AWS will kill any long-running processes (after 15 minutes). See [here](https://twitter.com/freekmurze/status/1306181930824667136) for an example.

Imagine you have a huge list of invoices in your DB, which need to be submitted one by one:

```php
class SendInvoices implements ShouldQueue
{
    use InteractsWithQueue, Queueable;

    public function handle()
    {
        Invoice::pending()->cursor()
            ->takeUntilTimeout(Carbon::createFromTimestamp(LARAVEL_START)->add(14, 'minutes'))
            ->each(fn ($invoice) => $invoice->submit());

        if (Invoice::pending()->exists()) {
            $this->release();
        }
    }
}
```

This will submit as many invoices as there is time for, and then dispatch the job again to keep going.

---

If you can submit invoices in bulk, you can even chunk them before the `takeUntilTimeout` call:

```php
Invoice::pending()->cursor()
    ->chunk(100)
    ->takeUntilTimeout(Carbon::createFromTimestamp(LARAVEL_START)->add(14, 'minutes'))
    ->each(fn ($invoices) => submit_invoices($invoices->all()));
```

---

**P.S.** This would've been _ssoooo_ much easier to test with `Carbon`, especially with the new time travel stuff. But it makes no sense to add `Carbon` as a dependency to `illuminate/collections` just for this.

I really hate all that stupid Mockery code I had to write, but I see no other way. Gives you a new level of appreciation for proper time travel :smile:

### CODE REVIEW

**Comment by clarkeash** [comment]
@JosephSilber does it make sense to use `DateTimeInterface` instead so people can also use `CarbonImmutable`/` DateTimeImmutable` as well?

**Comment by taylorotwell** [comment]
Yeah probably want `DateTimeInterface`.

**Comment by JosephSilber** [comment]
Done ✅

File: src/Illuminate/Collections/LazyCollection.php [hunk 1]
```diff
@@ -1385,4 +1401,14 @@ protected function passthru($method, array $params)
...
         });
     }
+
+    /**
+     * Get the current time.
+     *
+     * @return int
+     */
+    protected function now()
+    {
```
Comment (gocanto) [inline src/Illuminate/Collections/LazyCollection.php:1411]: would a setter/wither metod accepting the `datetime interface` be a better fit here? - hence, we would not need to allow for mockying protected methods.

Comment (JosephSilber) [reply src/Illuminate/Collections/LazyCollection.php:1411]: I'd rather have the complexity in the tests than in the implementation.

**Pull request merged**